### PR TITLE
Make sourceContext optional

### PR DESCRIPTION
**Pull request by komalali**
In cases where a self-hosted runner is being used, the source code may be baked into the runner. In those cases, an additional source step is unnecessary.

### CODE REVIEW

File: examples/yaml-deployment-settings/.pulumi
Comment (IaroslavTitov) [inline examples/yaml-deployment-settings/.pulumi:1]: What are the empty files for?

File: examples/examples_yaml_test.go [hunk 1]
```diff
@@ -208,6 +208,43 @@ func TestYamlDeploymentSettingsExample(t *testing.T) {
 	})
 }
 
+func TestYamlDeploymentSettingsNoSourceExample(t *testing.T) {
+
+	// Set up tmpdir with a Pulumi.yml with no resources
```
Comment (IaroslavTitov) [inline examples/examples_yaml_test.go:213]: How does this work? Is the ProgramTest running `up` on the folder with Pulumi.yml and then on a non-existing folder? Why is this needed?

Comment (komalali) [reply examples/examples_yaml_test.go:213]: This is copy-paste from the other TestYamlDeploymentSettingsExample above.

File: examples/yaml-deployment-settings/.pulumi
Comment (komalali) [reply examples/yaml-deployment-settings/.pulumi:1]: They're just being deleted, there was an issue with the integration test framework earlier where it was forcing us to add these empty files but the bug was fixed.

File: examples/examples_yaml_test.go (hunk 1 above)
Comment (IaroslavTitov) [reply examples/examples_yaml_test.go:213]: Duh, should've scrolled

Still curious as to why we need these temp dirs, but I can take a look later, unrelated to the change

**Review by IaroslavTitov — APPROVED** [review]

**Review by IaroslavTitov — APPROVED** [review]
